src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs

```diff
@@ -69,43 +69,50 @@ public void Serialize(object message, Stream stream)
 
         public object[] Deserialize(Stream stream, IList<Type> messageTypes)
         {
-            if (IsArrayStream(stream))
-            {
-                throw new Exception("Multiple messages in the same stream are not supported.");
-            }
+            var isArrayStream = IsArrayStream(stream);
 
             if (messageTypes.Any())
             {
-                return DeserializeMultipleMessageTypes(stream, messageTypes);
+                return DeserializeMultipleMessageTypes(stream, messageTypes, isArrayStream);
             }
 
+            return new[]
+            {
+                ReadObject(stream, isArrayStream, typeof(object))
+            };
+        }
+
+        object ReadObject(Stream stream, bool isArrayStream, Type type)
+        {
             using (var reader = readerCreator(stream))
             {
                 reader.CloseInput = false;
-                return new[]
+                if (isArrayStream)
                 {
-                    jsonSerializer.Deserialize<object>(reader)
-                };
+                    var objects = (object[])jsonSerializer.Deserialize(reader, type.MakeArrayType());
+                    if (objects.Length > 1)
+                    {
+                        throw new Exception("Multiple messages in the same stream are not supported.");
+                    }
+                    return objects.Single();
```

Comment:
Single or [0]? For me Single enforces one item being in the array, which is already checked for and is not necessary.